File: sdk/samples/07_advanced/python/dynamic_data.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0 OR MIT
# Copyright (c) 2025-2026 naskel.com

"""
Dynamic Data Sample - Demonstrates runtime type manipulation

Dynamic Data allows working with types at runtime without
compile-time type definitions. Useful for:
- Generic tools and data bridges
- Type discovery and introspection
- Protocol adapters and gateways

Key concepts:
- DynamicType: runtime type definition
- DynamicData: runtime data manipulation
- Type introspection
- Publishing and receiving dynamic data via HDDS

NOTE: CONCEPT DEMO - This sample demonstrates the APPLICATION PATTERN for DynamicData/DynamicType.
The native DynamicData/DynamicType API is not yet exported to the C/C++/Python SDK.
This sample uses standard participant/writer/reader API to show the concept.
"""

import os
import sys
import struct
import json
from dataclasses import dataclass, field
from enum import Enum, auto
from typing import Any, Dict, List, Optional
from copy import deepcopy

# Add SDK to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', '..', 'python'))

import hdds


class TypeKind(Enum):
    """Type kinds"""
    INT32 = auto()
    UINT32 = auto()
    INT64 = auto()
    FLOAT32 = auto()
    FLOAT64 = auto()
    BOOL = auto()
    STRING = auto()
    SEQUENCE = auto()
    ARRAY = auto()
    STRUCT = auto()
    UNION = auto()
    ENUM = auto()


def type_kind_str(kind: TypeKind) -> str:
    """Convert type kind to string"""
    return {
        TypeKind.INT32: "int32",
        TypeKind.UINT32: "uint32",
        TypeKind.INT64: "int64",
        TypeKind.FLOAT32: "float32",
        TypeKind.FLOAT64: "float64",
        TypeKind.BOOL: "bool",
        TypeKind.STRING: "string",
        TypeKind.SEQUENCE: "sequence",
        TypeKind.ARRAY: "array",
        TypeKind.STRUCT: "struct",
        TypeKind.UNION: "union",
        TypeKind.ENUM: "enum",
    }.get(kind, "unknown")


@dataclass
class MemberDescriptor:
    """Member descriptor"""
    name: str
    type: TypeKind
    id: int = 0
    is_key: bool = False
    is_optional: bool = False


class DynamicType:
    """Dynamic type definition"""

    def __init__(self, name: str, kind: TypeKind):
        self._name = name
        self._kind = kind
        self._members: List[MemberDescriptor] = []

    @property
    def name(self) -> str:
        return self._name

    @property
    def kind(self) -> TypeKind:
        return self._kind

    @property
    def members(self) -> List[MemberDescriptor]:
        return self._members

    def add_member(self, name: str, member_type: TypeKind,
                   is_key: bool = False, is_optional: bool = False):
        """Add a member to the type"""
        member = MemberDescriptor(
            name=name,
            type=member_type,
            id=len(self._members),
            is_key=is_key,
            is_optional=is_optional
        )
        self._members.append(member)

    def get_member(self, name: str) -> Optional[MemberDescriptor]:
        """Get member by name"""
        for m in self._members:
            if m.name == name:
                return m
        return None


@dataclass
class DataMember:
    """Dynamic data member"""
    name: str
    type: TypeKind
    value: Any = None
    is_set: bool = False


class DynamicData:
    """Dynamic data instance"""

    def __init__(self, dtype: DynamicType):
        self._type = dtype
        self._members: Dict[str, DataMember] = {}
        for m in dtype.members:
            self._members[m.name] = DataMember(
                name=m.name,
                type=m.type,
                value=None,
                is_set=False
            )

    @property
    def type(self) -> DynamicType:
        return self._type

    @property
    def members(self) -> Dict[str, DataMember]:
        return self._members

    # Setters
    def set_int32(self, name: str, value: int):
        if name in self._members:
            self._members[name].value = value
            self._members[name].is_set = True

    def set_float64(self, name: str, value: float):
        if name in self._members:
            self._members[name].value = value
            self._members[name].is_set = True

    def set_string(self, name: str, value: str):
        if name in self._members:
            self._members[name].value = value
            self._members[name].is_set = True

    def set_bool(self, name: str, value: bool):
        if name in self._members:
            self._members[name].value = value
            self._members[name].is_set = True

    # Getters
    def get_int32(self, name: str) -> int:
        if name in self._members and self._members[name].is_set:
            return self._members[name].value
        return 0

    def get_float64(self, name: str) -> float:
        if name in self._members and self._members[name].is_set:
            return self._members[name].value
        return 0.0

    def get_string(self, name: str) -> str:
        if name in self._members and self._members[name].is_set:
            return self._members[name].value
        return ""

    def get_bool(self, name: str) -> bool:
        if name in self._members and self._members[name].is_set:
            return self._members[name].value
        return False

    def clone(self) -> 'DynamicData':
        """Clone this dynamic data"""
        copy = DynamicData(self._type)
        copy._members = deepcopy(self._members)
        return copy

    def serialize(self) -> bytes:
        """Serialize to bytes using JSON encoding."""
        data = {
            "type": self._type.name,
            "members": {}
        }
        for name, member in self._members.items():
            if member.is_set:
                data["members"][name] = {
                    "type": type_kind_str(member.type),
                    "value": member.value
                }
        return json.dumps(data).encode('utf-8')

    @classmethod
    def deserialize(cls, data: bytes, dtype: DynamicType) -> 'DynamicData':
        """Deserialize from bytes."""
        parsed = json.loads(data.decode('utf-8'))
        result = cls(dtype)
        for name, member_data in parsed.get("members", {}).items():
            if name in result._members:
                result._members[name].value = member_data["value"]
                result._members[name].is_set = True
        return result


class TypeFactory:
    """Factory for creating dynamic types"""

    def __init__(self):
        self._types: Dict[str, DynamicType] = {}

    def create_struct(self, name: str) -> DynamicType:
        """Create a struct type"""
        dtype = DynamicType(name, TypeKind.STRUCT)
        self._types[name] = dtype
        return dtype

    def get_type(self, name: str) -> Optional[DynamicType]:
        """Get a type by name"""
        return self._types.get(name)


def print_type(dtype: DynamicType):
    """Print type information"""
    print(f"  Type: {dtype.name} ({type_kind_str(dtype.kind)})")
    print(f"  Members ({len(dtype.members)}):")
    for m in dtype.members:
        flags = ""
        if m.is_key:
            flags += " @key"
        if m.is_optional:
            flags += " @optional"
        print(f"    [{m.id}] {m.name}: {type_kind_str(m.type)}{flags}")


def print_data(data: DynamicData):
    """Print dynamic data"""
    print(f"  Data of type '{data.type.name}':")
    for name, member in data.members.items():
        if not member.is_set:
            print(f"    {name} = <unset>")
        elif member.type == TypeKind.STRING:
            print(f"    {name} = \"{member.value}\"")
        elif member.type == TypeKind.BOOL:
            print(f"    {name} = {'true' if member.value else 'false'}")
        else:
            print(f"    {name} = {member.value}")


def print_dynamic_data_overview():
    print("--- Dynamic Data Overview ---\n")
    print("Dynamic Data Architecture:\n")
    print("  +-----------------+      +-----------------+")
    print("  |  TypeFactory    |----->|  DynamicType    |")
    print("  |                 |      |  - name         |")
    print("  | create_struct() |      |  - kind         |")
    print("  | create_enum()   |      |  - members[]    |")
    print("  +-----------------+      +--------+--------+")
    print("                                    |")
    print("                                    v")
    print("                           +-----------------+")
    print("                           |  DynamicData    |")
    print("                           |  - type         |")
    print("                           |  - values[]     |")
    print("                           |  - get/set()    |")
    print("                           +-----------------+")
    print()
    print("Use Cases:")
    print("  - Generic data recording/replay tools")
    print("  - Protocol bridges (DDS <-> REST/MQTT)")
    print("  - Data visualization without type knowledge")
    print("  - Testing and debugging utilities")
    print()


def main():
    print("=== HDDS Dynamic Data Sample ===\n")
    print("NOTE: CONCEPT DEMO - Native DynamicData/DynamicType API not yet in SDK.")
    print("      Using standard pub/sub API to demonstrate the pattern.\n")

    print_dynamic_data_overview()

    # Initialize HDDS logging
    hdds.logging.init(hdds.LogLevel.INFO)

    # Create type factory
    factory = TypeFactory()
    print("[OK] TypeFactory created\n")

    # Define a SensorReading type at runtime
    print("--- Creating Dynamic Type ---\n")

    sensor_type = factory.create_struct("SensorReading")
    sensor_type.add_member("sensor_id", TypeKind.INT32, is_key=True)
    sensor_type.add_member("location", TypeKind.STRING)
    sensor_type.add_member("temperature", TypeKind.FLOAT64)
    sensor_type.add_member("humidity", TypeKind.FLOAT64)
    sensor_type.add_member("is_valid", TypeKind.BOOL)

    print("[OK] Type 'SensorReading' created dynamically\n")
    print_type(sensor_type)
    print()

    # Create and populate dynamic data
    print("--- Creating Dynamic Data ---\n")

    reading1 = DynamicData(sensor_type)
    reading1.set_int32("sensor_id", 101)
    reading1.set_string("location", "Building-A/Room-1")
    reading1.set_float64("temperature", 23.5)
    reading1.set_float64("humidity", 45.2)
    reading1.set_bool("is_valid", True)

    print("[OK] DynamicData instance created\n")
    print_data(reading1)
    print()

    # Read values back
    print("--- Reading Dynamic Data ---\n")

    id_val = reading1.get_int32("sensor_id")
    loc = reading1.get_string("location")
    temp = reading1.get_float64("temperature")
    hum = reading1.get_float64("humidity")
    valid = reading1.get_bool("is_valid")

    print("Read values:")
    print(f"  sensor_id: {id_val}")
    print(f"  location: {loc}")
    print(f"  temperature: {temp}")
    print(f"  humidity: {hum}")
    print(f"  is_valid: {'true' if valid else 'false'}\n")

    # Publish dynamic data via HDDS
    print("--- Publishing Dynamic Data via HDDS ---\n")

    participant = hdds.Participant("DynamicDataDemo")
    writer = participant.create_writer("DynamicSensorTopic", qos=hdds.QoS.reliable())
    reader = participant.create_reader("DynamicSensorTopic", qos=hdds.QoS.reliable())

    print("[OK] Created participant, writer, and reader")

    # Serialize and publish
    serialized = reading1.serialize()
    writer.write(serialized)
    print(f"[OK] Published dynamic data ({len(serialized)} bytes)")

    # Wait briefly for data to arrive
    import time
    time.sleep(0.1)

    # Read and deserialize
    received_data = reader.take()
    if received_data:
        received = DynamicData.deserialize(received_data, sensor_type)
        print("[OK] Received and deserialized dynamic data\n")
        print_data(received)
    else:
        print("[INFO] No data received (this is normal in single-process demo)")
    print()

    # Clone data
    print("--- Cloning Dynamic Data ---\n")

    reading2 = reading1.clone()
    reading2.set_int32("sensor_id", 102)
    reading2.set_string("location", "Building-B/Room-3")
    reading2.set_float64("temperature", 25.0)

    print("[OK] Cloned and modified:\n")
    print_data(reading2)
    print()

    # Type introspection
    print("--- Type Introspection ---\n")

    print("Iterating over type members:")
    for m in sensor_type.members:
        print(f"  Member '{m.name}':")
        print(f"    - Type: {type_kind_str(m.type)}")
        print(f"    - ID: {m.id}")
        print(f"    - Is key: {'yes' if m.is_key else 'no'}")
        print(f"    - Optional: {'yes' if m.is_optional else 'no'}")
    print()

    # Create another type
    print("--- Creating Additional Type ---\n")

    alarm_type = factory.create_struct("AlarmEvent")
    alarm_type.add_member("alarm_id", TypeKind.INT32, is_key=True)
    alarm_type.add_member("severity", TypeKind.INT32)
    alarm_type.add_member("message", TypeKind.STRING)
    alarm_type.add_member("acknowledged", TypeKind.BOOL)

    print_type(alarm_type)
    print()

    alarm = DynamicData(alarm_type)
    alarm.set_int32("alarm_id", 5001)
    alarm.set_int32("severity", 3)
    alarm.set_string("message", "High temperature warning")
    alarm.set_bool("acknowledged", False)

    print_data(alarm)
    print()

    # Publish alarm via HDDS
    alarm_writer = participant.create_writer("AlarmTopic", qos=hdds.QoS.reliable())
    alarm_writer.write(alarm.serialize())
    print("[OK] Published alarm event via HDDS\n")

    # Best practices
    print("--- Dynamic Data Best Practices ---\n")
    print("1. Cache type lookups for performance-critical paths")
    print("2. Use member IDs instead of names for faster access")
    print("3. Validate type compatibility before operations")
    print("4. Consider memory management for string members")
    print("5. Use typed APIs when types are known at compile time")
    print("6. Leverage type introspection for generic tooling")

    print("\n=== Sample Complete ===")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: sdk/samples/07_advanced/cpp/dynamic_data.cpp
// SPDX-License-Identifier: Apache-2.0 OR MIT
// Copyright (c) 2025-2026 naskel.com

/**
 * Dynamic Data Sample - Demonstrates runtime type manipulation
 *
 * Dynamic Data allows working with types at runtime without
 * compile-time type definitions. Useful for:
 * - Generic tools and data bridges
 * - Type discovery and introspection
 * - Protocol adapters and gateways
 *
 * Key concepts:
 * - DynamicType: runtime type definition
 * - DynamicData: runtime data manipulation
 * - Type introspection
 *
 * Uses the real HDDS C++ API for pub/sub transport with
 * application-level dynamic data representation.
 *
 * NOTE: CONCEPT DEMO - This sample demonstrates the APPLICATION PATTERN for DynamicData/DynamicType.
 * The native DynamicData/DynamicType API is not yet exported to the C/C++/Python SDK.
 * This sample uses standard participant/writer/reader API to show the concept.
 */

#include <hdds.hpp>
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <variant>
#include <optional>
#include <memory>
#include <cstring>
#include <thread>
#include <chrono>

#include "generated/HelloWorld.hpp"

using namespace hdds_samples;
using namespace std::chrono_literals;

// Type kinds
enum class TypeKind {
    Int32,
    UInt32,
    Int64,
    Float32,
    Float64,
    Bool,
    String,
    Sequence,
    Array,
    Struct,
};

std::string type_kind_str(TypeKind kind) {
    switch (kind) {
        case TypeKind::Int32: return "int32";
        case TypeKind::UInt32: return "uint32";
        case TypeKind::Int64: return "int64";
        case TypeKind::Float32: return "float32";
        case TypeKind::Float64: return "float64";
        case TypeKind::Bool: return "bool";
        case TypeKind::String: return "string";
        case TypeKind::Sequence: return "sequence";
        case TypeKind::Array: return "array";
        case TypeKind::Struct: return "struct";
        default: return "unknown";
    }
}

// Member descriptor
struct MemberDescriptor {
    std::string name;
    TypeKind type;
    uint32_t id;
    bool is_key = false;
    bool is_optional = false;
};

// Dynamic type definition
class DynamicType {
public:
    DynamicType(const std::string& name, TypeKind kind)
        : name_(name), kind_(kind) {}

    void add_member(const std::string& name, TypeKind type,
                    bool is_key = false, bool is_optional = false) {
        MemberDescriptor member;
        member.name = name;
        member.type = type;
        member.id = static_cast<uint32_t>(members_.size());
        member.is_key = is_key;
        member.is_optional = is_optional;
        members_.push_back(member);
    }

    const std::string& name() const { return name_; }
    TypeKind kind() const { return kind_; }
    const std::vector<MemberDescriptor>& members() const { return members_; }

    std::optional<MemberDescriptor> get_member(const std::string& name) const {
        for (const auto& m : members_) {
            if (m.name == name) return m;
        }
        return std::nullopt;
    }

private:
    std::string name_;
    TypeKind kind_;
    std::vector<MemberDescriptor> members_;
};

// Dynamic data value type
using DataValue = std::variant<
    int32_t, uint32_t, int64_t,
    float, double, bool, std::string,
    std::monostate  // for unset values
>;

// Dynamic data implementation
class DynamicData {
public:
    explicit DynamicData(std::shared_ptr<DynamicType> type)
        : type_(type) {
        for (const auto& m : type_->members()) {
            values_[m.name] = std::monostate{};
            is_set_[m.name] = false;
        }
    }

    // Setters
    void set_int32(const std::string& name, int32_t value) {
        if (values_.count(name)) { values_[name] = value; is_set_[name] = true; }
    }

    void set_float64(const std::string& name, double value) {
        if (values_.count(name)) { values_[name] = value; is_set_[name] = true; }
    }

    void set_string(const std::string& name, const std::string& value) {
        if (values_.count(name)) { values_[name] = value; is_set_[name] = true; }
    }

    void set_bool(const std::string& name, bool value) {
        if (values_.count(name)) { values_[name] = value; is_set_[name] = true; }
    }

    // Getters
    int32_t get_int32(const std::string& name) const {
        if (auto it = values_.find(name); it != values_.end()) {
            if (auto* v = std::get_if<int32_t>(&it->second)) return *v;
        }
        return 0;
    }

    double get_float64(const std::string& name) const {
        if (auto it = values_.find(name); it != values_.end()) {
            if (auto* v = std::get_if<double>(&it->second)) return *v;
        }
        return 0.0;
    }

    std::string get_string(const std::string& name) const {
        if (auto it = values_.find(name); it != values_.end()) {
            if (auto* v = std::get_if<std::string>(&it->second)) return *v;
        }
        return "";
    }

    bool get_bool(const std::string& name) const {
        if (auto it = values_.find(name); it != values_.end()) {
            if (auto* v = std::get_if<bool>(&it->second)) return *v;
        }
        return false;
    }

    bool is_set(const std::string& name) const {
        auto it = is_set_.find(name);
        return it != is_set_.end() && it->second;
    }

    // Serialize to bytes for transmission
    std::vector<uint8_t> serialize() const {
        // Simple serialization format: [member_count][member_data...]
        std::vector<uint8_t> data;
        uint32_t count = static_cast<uint32_t>(type_->members().size());
        data.resize(4);
        std::memcpy(data.data(), &count, 4);

        for (const auto& m : type_->members()) {
            auto it = values_.find(m.name);
            if (it == values_.end()) continue;

            std::visit([&data, &m](auto&& val) {
                using T = std::decay_t<decltype(val)>;
                if constexpr (std::is_same_v<T, int32_t>) {
                    size_t pos = data.size();
                    data.resize(pos + 4);
                    std::memcpy(data.data() + pos, &val, 4);
                } else if constexpr (std::is_same_v<T, double>) {
                    size_t pos = data.size();
                    data.resize(pos + 8);
                    std::memcpy(data.data() + pos, &val, 8);
                } else if constexpr (std::is_same_v<T, std::string>) {
                    uint32_t len = static_cast<uint32_t>(val.size());
                    size_t pos = data.size();
                    data.resize(pos + 4 + len);
                    std::memcpy(data.data() + pos, &len, 4);
                    std::memcpy(data.data() + pos + 4, val.c_str(), len);
                } else if constexpr (std::is_same_v<T, bool>) {
                    data.push_back(val ? 1 : 0);
                }
            }, it->second);
        }
        return data;
    }

    // Clone
    std::unique_ptr<DynamicData> clone() const {
        auto copy = std::make_unique<DynamicData>(type_);
        copy->values_ = values_;
        copy->is_set_ = is_set_;
        return copy;
    }

    std::shared_ptr<DynamicType> type() const { return type_; }
    const std::map<std::string, DataValue>& values() const { return values_; }

private:
    std::shared_ptr<DynamicType> type_;
    std::map<std::string, DataValue> values_;
    std::map<std::string, bool> is_set_;
};

// Type factory
class TypeFactory {
public:
    std::shared_ptr<DynamicType> create_struct(const std::string& name) {
        auto type = std::make_shared<DynamicType>(name, TypeKind::Struct);
        types_[name] = type;
        return type;
    }

    std::shared_ptr<DynamicType> get_type(const std::string& name) const {
        if (auto it = types_.find(name); it != types_.end()) {
            return it->second;
        }
        return nullptr;
    }

private:
    std::map<std::string, std::shared_ptr<DynamicType>> types_;
};

void print_type(const DynamicType& type) {
    std::cout << "  Type: " << type.name() << " (" << type_kind_str(type.kind()) << ")\n";
    std::cout << "  Members (" << type.members().size() << "):\n";
    for (const auto& m : type.members()) {
        std::cout << "    [" << m.id << "] " << m.name << ": " << type_kind_str(m.type);
        if (m.is_key) std::cout << " @key";
        if (m.is_optional) std::cout << " @optional";
        std::cout << "\n";
    }
}

void print_data(const DynamicData& data) {
    std::cout << "  Data of type '" << data.type()->name() << "':\n";
    for (const auto& m : data.type()->members()) {
        std::cout << "    " << m.name << " = ";
        if (!data.is_set(m.name)) {
            std::cout << "<unset>\n";
            continue;
        }

        auto it = data.values().find(m.name);
        if (it == data.values().end()) {
            std::cout << "<not found>\n";
            continue;
        }

        std::visit([](auto&& val) {
            using T = std::decay_t<decltype(val)>;
            if constexpr (std::is_same_v<T, std::monostate>) {
                std::cout << "<unset>";
            } else if constexpr (std::is_same_v<T, std::string>) {
                std::cout << "\"" << val << "\"";
            } else if constexpr (std::is_same_v<T, bool>) {
                std::cout << (val ? "true" : "false");
            } else {
                std::cout << val;
            }
        }, it->second);
        std::cout << "\n";
    }
}

void print_dynamic_data_overview() {
    std::cout << "--- Dynamic Data Overview ---\n\n";
    std::cout << "Dynamic Data Architecture:\n\n";
    std::cout << "  +------------------+      +------------------+\n";
    std::cout << "  |  TypeFactory     |----->|  DynamicType     |\n";
    std::cout << "  |                  |      |  - name          |\n";
    std::cout << "  | create_struct()  |      |  - kind          |\n";
    std::cout << "  +------------------+      |  - members[]     |\n";
    std::cout << "                           +--------+---------+\n";
    std::cout << "                                    |\n";
    std::cout << "                                    v\n";
    std::cout << "                           +------------------+\n";
    std::cout << "                           |  DynamicData     |\n";
    std::cout << "                           |  - type          |\n";
    std::cout << "                           |  - values[]      |\n";
    std::cout << "                           |  - get/set()     |\n";
    std::cout << "                           +------------------+\n";
    std::cout << "\n";
    std::cout << "Use Cases:\n";
    std::cout << "  - Generic data recording/replay tools\n";
    std::cout << "  - Protocol bridges (DDS <-> REST/MQTT)\n";
    std::cout << "  - Data visualization without type knowledge\n";
    std::cout << "  - Testing and debugging utilities\n";
    std::cout << "\n";
}

void run_publisher(hdds::Participant& participant, std::shared_ptr<DynamicType> sensor_type) {
    std::cout << "--- Publisher Mode ---\n\n";

    auto writer = participant.create_writer_raw("DynamicDataTopic", hdds::QoS::reliable());
    std::cout << "[OK] Writer created for DynamicDataTopic\n\n";

    // Publish dynamic data samples
    for (int i = 0; i < 5; i++) {
        DynamicData reading(sensor_type);
        reading.set_int32("sensor_id", 100 + i);
        reading.set_string("location", "Building-A/Room-" + std::to_string(i + 1));
        reading.set_float64("temperature", 20.0 + i * 1.5);
        reading.set_float64("humidity", 45.0 + i * 2.0);
        reading.set_bool("is_valid", true);

        std::cout << "Publishing:\n";
        print_data(reading);
        std::cout << "\n";

        auto bytes = reading.serialize();
        writer->write_raw(bytes);

        std::this_thread::sleep_for(500ms);
    }

    std::cout << "Done publishing.\n";
}

void run_subscriber(hdds::Participant& participant) {
    std::cout << "--- Subscriber Mode ---\n\n";

    auto reader = participant.create_reader_raw("DynamicDataTopic", hdds::QoS::reliable());
    std::cout << "[OK] Reader created for DynamicDataTopic\n";

    hdds::WaitSet waitset;
    waitset.attach(reader->get_status_condition());
    std::cout << "[OK] WaitSet attached\n\n";

    std::cout << "Waiting for dynamic data samples...\n\n";

    int received = 0;
    int timeout_count = 0;

    while (received < 5 && timeout_count < 5) {
        if (waitset.wait(2s)) {
            while (auto sample = reader->take_raw()) {
                std::cout << "[Received] Raw bytes: " << sample->size() << " bytes\n";
                received++;
            }
            timeout_count = 0;
        } else {
            timeout_count++;
            std::cout << "(waiting...)\n";
        }
    }

    std::cout << "\nReceived " << received << " samples.\n";
}

int main(int argc, char* argv[]) {
    std::cout << "=== HDDS Dynamic Data Sample ===\n\n";
    std::cout << "NOTE: CONCEPT DEMO - Native DynamicData/DynamicType API not yet in SDK.\n"
              << "      Using standard pub/sub API to demonstrate the pattern.\n\n";

    bool is_publisher = (argc > 1) &&
        (std::strcmp(argv[1], "pub") == 0 ||
         std::strcmp(argv[1], "publisher") == 0 ||
         std::strcmp(argv[1], "-p") == 0);

    print_dynamic_data_overview();

    try {
        // Initialize logging
        hdds::logging::init(hdds::LogLevel::Warn);

        // Create type factory
        TypeFactory factory;
        std::cout << "[OK] TypeFactory created\n\n";

        // Define a SensorReading type at runtime
        std::cout << "--- Creating Dynamic Type ---\n\n";

        auto sensor_type = factory.create_struct("SensorReading");
        sensor_type->add_member("sensor_id", TypeKind::Int32, true, false);
        sensor_type->add_member("location", TypeKind::String, false, false);
        sensor_type->add_member("temperature", TypeKind::Float64, false, false);
        sensor_type->add_member("humidity", TypeKind::Float64, false, false);
        sensor_type->add_member("is_valid", TypeKind::Bool, false, false);

        std::cout << "[OK] Type 'SensorReading' created dynamically\n\n";
        print_type(*sensor_type);
        std::cout << "\n";

        // Create and populate dynamic data
        std::cout << "--- Creating Dynamic Data ---\n\n";

        DynamicData reading1(sensor_type);
        reading1.set_int32("sensor_id", 101);
        reading1.set_string("location", "Building-A/Room-1");
        reading1.set_float64("temperature", 23.5);
        reading1.set_float64("humidity", 45.2);
        reading1.set_bool("is_valid", true);

        std::cout << "[OK] DynamicData instance created\n\n";
        print_data(reading1);
        std::cout << "\n";

        // Read values back
        std::cout << "--- Reading Dynamic Data ---\n\n";

        auto id = reading1.get_int32("sensor_id");
        auto loc = reading1.get_string("location");
        auto temp = reading1.get_float64("temperature");
        auto hum = reading1.get_float64("humidity");
        auto valid = reading1.get_bool("is_valid");

        std::cout << "Read values:\n";
        std::cout << "  sensor_id: " << id << "\n";
        std::cout << "  location: " << loc << "\n";
        std::cout << "  temperature: " << temp << "\n";
        std::cout << "  humidity: " << hum << "\n";
        std::cout << "  is_valid: " << (valid ? "true" : "false") << "\n\n";

        // Clone data
        std::cout << "--- Cloning Dynamic Data ---\n\n";

        auto reading2 = reading1.clone();
        reading2->set_int32("sensor_id", 102);
        reading2->set_string("location", "Building-B/Room-3");
        reading2->set_float64("temperature", 25.0);

        std::cout << "[OK] Cloned and modified:\n\n";
        print_data(*reading2);
        std::cout << "\n";

        // Type introspection
        std::cout << "--- Type Introspection ---\n\n";

        std::cout << "Iterating over type members:\n";
        for (const auto& m : sensor_type->members()) {
            std::cout << "  Member '" << m.name << "':\n";
            std::cout << "    - Type: " << type_kind_str(m.type) << "\n";
            std::cout << "    - ID: " << m.id << "\n";
            std::cout << "    - Is key: " << (m.is_key ? "yes" : "no") << "\n";
            std::cout << "    - Optional: " << (m.is_optional ? "yes" : "no") << "\n";
        }
        std::cout << "\n";

        // Create participant and demonstrate pub/sub with dynamic data
        std::cout << "--- HDDS Pub/Sub with Dynamic Data ---\n\n";

        hdds::Participant participant("DynamicDataDemo");
        std::cout << "[OK] Participant created\n\n";

        if (is_publisher) {
            run_publisher(participant, sensor_type);
        } else {
            run_subscriber(participant);
        }

        // Best practices
        std::cout << "\n--- Dynamic Data Best Practices ---\n\n";
        std::cout << "1. Cache type lookups for performance-critical paths\n";
        std::cout << "2. Use member IDs instead of names for faster access\n";
        std::cout << "3. Validate type compatibility before operations\n";
        std::cout << "4. Consider memory management for string members\n";
        std::cout << "5. Use typed APIs when types are known at compile time\n";
        std::cout << "6. Leverage type introspection for generic tooling\n";

        std::cout << "\n=== Sample Complete ===\n";

    } catch (const hdds::Error& e) {
        std::cerr << "HDDS Error: " << e.what() << std::endl;
        return 1;
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}
